pandas/core/tools/datetimes.py

```diff
@@ -305,6 +310,51 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
             except (ValueError, TypeError):
                 raise e
 
+    def _maybe_convert_cache(arg, cache, box, format, name=None, tz=tz):
+        """
+        Try to convert the datetimelike arg using
+        a cache of converted dates.
+
+        Parameters
+        ----------
+        arg : integer, float, string, datetime, list, tuple, 1-d array, Series
+            Datetime argument to convert     
+        cache : boolean
+            If True, try to convert the dates with a cache
+            If False, short circuit and return None
+            Flag whether to cache the converted dates
+        box : boolean
+            If True, return a DatetimeIndex  
+            if False, return an ndarray of values     
+        tz : String or None
+            'utc' if UTC=True was passed else None
+        name : String, default None
+            DatetimeIndex name
+        Returns
+        -------
+        Series if original argument was a Series
+        DatetimeIndex if box=True and original argument was not a Series
+        ndarray if box=False and original argument was not a Series
+        None if the conversion failed
+        """
+        if cache and is_list_like(arg) and len(arg) >= 1000:
+            # Perform a quicker unique check
+            from pandas import Index
+            if not Index(arg).is_unique:
+                unique_dates = algorithms.unique(arg)
+                from pandas import Series
+                cache_dates = _convert_listlike(unique_dates, True, format,
+                                                tz=tz)
+                convert_cache = Series(cache_dates, index=unique_dates)
+                result = Series(arg, name=name).map(convert_cache)
+                if isinstance(arg, Series):
+                    return result
+                elif box:
```

Comment:
if u use _convert_list_like where u return None and do the boxing outside (iow where this is called) then you can make this simpler and avoid splitting logic 

now it’s ok 2 places

Reply:
Good call on return _convert_list_like instead of None.

However if I move the boxing outside this function, I think the code will be less DRY since I'll have to repeat the boxing logic twice for Indexes and list-like arguments. So essentially if:

```
def _maybe_convert_cache(...):
    if cache:
        ...
        return Series(arg, name=name).map(convert_cache)
    else: 
        return _convert_cache(...)
```
then the logic further down will be:

```
elif isinstance(arg, ABCIndexClass):
    result = _maybe_convert_cache(...)
    if box:
        return DatetimeIndex(...)
    else:
        return result.values
elif is_list_like(arg):
    result = _maybe_convert_cache(...)
    if box:
        return DatetimeIndex(...)
    else:
        return result.values
```

Reply:
see my comments above, this needs re-factoring.